### DISCUSSION TITLE
depends-on-app 依赖 helm 类型workload 引发的cue: marshal  error: load.apply.value: undefined field "data"

### DISCUSSION THREAD

**Discussion by ranger-huang** [kubevela/kubevela#3340] [Q&A] [answered]
原有应用 `minio` 有类型为 helm、名为  `minio` 的应用组件， 而应用组件`minio`中会渲染出同名的configmap `minio` ; 
后来添加新应用 demo-app，依赖 应用`minio`，在执行到 depends-minio 步骤时报错如下：

```
$> kubectl get app -oyaml demo-app
spec:
  workflow:
    steps:
       - name: depends-minio
          type: depends-on-app
          properties:
            name: minio
            namespace: default
status:
  workflow:
    steps:
     - firstExecuteTime: '2022-02-28T06:28:42Z'
        id: 579rvzsq6h
        lastExecuteTime: '2022-02-28T06:29:35Z'
        message: >-
          step load: step apply: run step(provider=kube,do=apply): cue: marshal
          error: load.apply.value: undefined field "data"
        name: depends-minio
        phase: failed
        reason: Execute
        type: depends-on-app
...
```

初步定位到的问题出现在以下地方。

当应用 `minio`的helm release渲染后单应用尚未稳定，此处 `dependsOn.err != _|_`  内 apply 了原应用 `minio` 同名configmap 中的data['application'] ，实际上minio helm release中的 configmap 并没有data['application'] ，因此会报出以上错误。

https://github.com/oam-dev/kubevela/blob/5b59db5f0bd759451937f157c28ab34bfa817f1a/charts/vela-core/templates/defwithtemplate/depends-on-app.yaml#L21-L50

这部分处理逻辑，应该是认为 depends-on-app 内的应用， 是 vela addon 应用，因此与应用同名的应用组件 helm  `minio` 引起了冲突。

不知道这算不算是一个问题，depends-on-app 在 L32 `dependsOn.err != _|_`  之后的处理 ，是不是应该考虑避免 vela addon 和 同名应用和应用组件 产生冲突？

**Comment by FogDong** [comment, marked as answer]
感谢提出！可以考虑修改 cue 内的逻辑，对 configmap 中没有 application 的情况做处理，如果 configmap 中没有 application，就走原来的逻辑。不知道您有没有兴趣贡献修复一下？
